hack/generator/pkg/astmodel/definition.go

```diff
@@ -5,58 +5,45 @@
-	RequiredImports() []PackageReference
-}
-
-// ReferenceChecker is used to check for references to a specific definition
-type ReferenceChecker interface {
-	// References determines if this type has a direct reference to the given definition name
-	// For example, a struct references its field
-	References(d *DefinitionName) bool
-}
-
-// HasRelatedDefinitions specifies that the type might create additional supporting definitions
-type HasRelatedDefinitions interface {
-	// CreateRelatedDefinitions returns any additional definitions related to this one
-	// (This allows one definition to act as a factory creating others)
-	CreateRelatedDefinitions(ref PackageReference, namehint string, idFactory IdentifierFactory) []Definition
-}
+// TypeDefiner represents a named type in the output files, and knows how to generate the Go AST
+type TypeDefiner interface {
 
-// Definition represents models that can render into Go code
-type Definition interface {
-	HasRelatedDefinitions
+	// Name is the name that will be bound to the type
+	Name() *TypeName
 
-	// FileNameHint returns what a file that contains this definition (if any) should be called
-	// this is not always used as we might combine multiple definitions into one file
-	FileNameHint() string
+	// Type is the type that the name will be bound to
+	Type() Type
 
-	// AsDecalarations() generates the Go code representing this definition
+	// AsDeclarations generates the actual Go declarations
 	AsDeclarations() []ast.Decl
+}
 
-	// DefinitionName: How do you refer to this definition?
-	Reference() *DefinitionName
-
-	// Type: What is the type associated to this definition?
-	Type() Type
-
-	// Tidy cleans up the definition prior to code generation
-	Tidy()
+// FileNameHint returns what a file that contains this definition (if any) should be called
+// this is not always used as we might combine multiple definitions into one file
+func FileNameHint(def TypeDefiner) string {
+	return def.Name().name
 }
 
 // Type represents something that is a Go type
 type Type interface {
-	HasImports
-	ReferenceChecker
-	HasRelatedDefinitions
+	// RequiredImports returns a list of packages required by this type
+	RequiredImports() []PackageReference
+
+	// References determines if this type has a direct reference to the given definition name
+	// For example an Array of Persons references a Person
+	References(d *TypeName) bool
 
-	// AsType renders the current instance as a Go abstract syntax tree
+	// AsType renders as a Go abstract syntax tree for a type
+	// (yes this says ast.Expr but that is what the Go 'ast' package uses for types)
 	AsType() ast.Expr
 
 	// Equals returns true if the passed type is the same as this one, false otherwise
 	Equals(t Type) bool
+
+	// CreateDefinitions gives a name to the type and might generate some asssociated definitions as well (the second result)
+	CreateDefinitions(name *TypeName, idFactory IdentifierFactory) (TypeDefiner, []TypeDefiner)
```

Comment:
one is the actual type being defined, so I thought it was more obvious to make that clear rather than return a `[]TypeDefiner` which we then have to slice into parts